File: frontend/src/utils/permission.js
import axios from 'axios'
import { MENU_STRUCTURE } from '@/config/menu'

// 权限管理工具类
class PermissionManager {
  constructor() {
    this.permissions = []
    this.userRole = ''
    this.userRoleCode = ''
    this.isInitialized = false
  }

  // 初始化权限管理器
  async init() {
    try {
      const response = await axios.get('/api/v1/permissions/user/menu')
      if (response.data.code === 0) {
        this.permissions = response.data.data.permissions || []
        this.userRole = response.data.data.user_role || ''
        this.userRoleCode = response.data.data.user_role_code || ''
        this.isInitialized = true
        console.log('权限初始化成功:', this.permissions.length, '个权限')
      } else {
        console.error('权限初始化失败:', response.data.message)
      }
    } catch (error) {
      console.error('权限初始化错误:', error)
      // 如果初始化失败，使用默认权限（所有菜单可见）
      this.isInitialized = true
    }
  }

  // 检查用户是否有某个权限
  hasPermission(permissionCode) {
    if (!this.isInitialized) {
      console.warn('权限管理器未初始化，默认允许访问')
      return true
    }

    // 超级管理员拥有所有权限
    if (this.userRoleCode === 'super_admin') {
      return true
    }

    // 检查权限列表
    return this.permissions.some(permission => 
      permission.permission_code === permissionCode && permission.status === 1
    )
  }

  // 检查用户是否有任意一个权限
  hasAnyPermission(permissionCodes) {
    if (!this.isInitialized) {
      return true
    }

    if (this.userRoleCode === 'super_admin') {
      return true
    }

    return permissionCodes.some(code => this.hasPermission(code))
  }

  // 获取用户角色信息
  getUserRole() {
    return {
      role: this.userRole,
      roleCode: this.userRoleCode
    }
  }

  // 获取用户的菜单权限
  getMenuPermissions() {
    return this.permissions.filter(permission => 
      permission.resource_type === 'menu' || permission.resource_type === 'module'
    )
  }

  // 根据权限生成菜单结构
  generateMenuStructure() {
    // 使用统一的菜单结构配置
    const filteredMenus = []
    
    MENU_STRUCTURE.forEach(menuItem => {
      // 检查主菜单权限
      if (this.hasPermission(menuItem.permission)) {
        // 过滤子菜单
        const filteredChildren = menuItem.children ? 
          menuItem.children.filter(child => this.hasPermission(child.permission)) : []
        
        // 如果有子菜单或者主菜单本身有独立路径，则保留
        if (filteredChildren.length > 0 || menuItem.path) {
          filteredMenus.push({
            ...menuItem,
            children: filteredChildren
          })
        }
      }
    })

    return filteredMenus
  }

  // 重置权限管理器
  reset() {
    this.permissions = []
    this.userRole = ''
    this.userRoleCode = ''
    this.isInitialized = false
  }
}

// 创建全局权限管理器实例
const permissionManager = new PermissionManager()

export default permissionManager
=======
  // 重置权限管理器
  reset() {
    this.permissions = []
    this.userRole = ''
    this.userRoleCode = ''
    this.isInitialized = false
  }
}

// 创建全局权限管理器实例
const permissionManager = new PermissionManager()

export default permissionManager
      dashboard: {
        name: '仪表板',
        path: '/',
        icon: 'Odometer',
        permission: 'dashboard:view',
        children: []
      },
      stocks: {
        name: '股票数据',
        path: '/stocks',
        icon: 'DataLine',
        permission: 'stocks:module',
        children: [
          { name: '股票列表', path: '/stocks', permission: 'stocks:list' },
          { name: '自选股', path: '/stocks/watchlist', permission: 'stocks:watchlist' },
          { name: '实时行情', path: '/stocks/realtime', permission: 'stocks:realtime' }
        ]
      },
      strategies: {
        name: '策略管理',
        path: '/strategies',
        icon: 'SetUp',
        permission: 'strategies:module',
        children: [
          { name: '策略列表', path: '/strategies', permission: 'strategies:list' },
          { name: '创建策略', path: '/strategies/create', permission: 'strategies:create' },
          { name: '回测中心', path: '/strategies/backtest', permission: 'strategies:backtest' }
        ]
      },
      selections: {
        name: '选股结果',
        path: '/selections',
        icon: 'Select',
        permission: 'selections:module',
        children: [
          { name: '选股记录', path: '/selections', permission: 'selections:list' },
          { name: '今日选股', path: '/selections/today', permission: 'selections:today' },
          { name: '历史回顾', path: '/selections/history', permission: 'selections:history' }
        ]
      },
      portfolios: {
        name: '投资组合',
        path: '/portfolios',
        icon: 'PieChart',
        permission: 'portfolios:module',
        children: [
          { name: '组合管理', path: '/portfolios', permission: 'portfolios:manage' },
          { name: '业绩分析', path: '/portfolios/performance', permission: 'portfolios:performance' }
        ]
      },
      collectors: {
        name: '数据采集',
        path: '/collectors',
        icon: 'Download',
        permission: 'collectors:module',
        children: [
          { name: '采集器管理', path: '/collectors', permission: 'collectors:manage' },
          { name: '采集任务', path: '/collectors/tasks', permission: 'collectors:tasks' },
          { name: '数据同步', path: '/collectors/sync', permission: 'collectors:sync' }
        ]
      },
      notifications: {
        name: '通知管理',
        path: '/notifications/robots',
        icon: 'Bell',
        permission: 'notifications:module',
        children: [
          { name: '机器人配置', path: '/notifications/robots', permission: 'notifications:robots' },
          { name: '消息模板', path: '/notifications/templates', permission: 'notifications:templates' },
          { name: '发送日志', path: '/notifications/logs', permission: 'notifications:logs' }
        ]
      },
      reports: {
        name: '报表分析',
        path: '/reports/strategy',
        icon: 'Document',
        permission: 'reports:module',
        children: [
          { name: '策略报告', path: '/reports/strategy', permission: 'reports:strategy' },
          { name: '市场分析', path: '/reports/market', permission: 'reports:market' },
          { name: '风险分析', path: '/reports/risk', permission: 'reports:risk' }
        ]
      },
      system: {
        name: '系统管理',
        path: '/system/users',
        icon: 'Setting',
        permission: 'system:module',
        children: [
          { name: '用户管理', path: '/system/users', permission: 'system:users' },
          { name: '角色管理', path: '/system/roles', permission: 'system:roles' },
          { name: '系统监控', path: '/system/monitoring', permission: 'system:monitoring' },
          { name: '系统设置', path: '/system/settings', permission: 'system:settings' }
        ]
      }
    }

    // 过滤没有权限的菜单项
    const filteredMenu = {}
    
    Object.keys(menuTree).forEach(key => {
      const menuItem = menuTree[key]
      
      // 检查主菜单权限
      if (this.hasPermission(menuItem.permission)) {
        // 过滤子菜单
        const filteredChildren = menuItem.children.filter(child => 
          this.hasPermission(child.permission)
        )
        
        // 如果有子菜单或者主菜单本身有独立路径，则保留
        if (filteredChildren.length > 0 || menuItem.path) {
          filteredMenu[key] = {
            ...menuItem,
            children: filteredChildren
          }
        }
      }
    })

    return filteredMenu
  }

  // 重置权限管理器
  reset() {
    this.permissions = []
    this.userRole = ''
    this.userRoleCode = ''
    this.isInitialized = false
  }
}

// 创建全局权限管理器实例
const permissionManager = new PermissionManager()

export default permissionManager

File: internal/middleware/permission.go
package middleware

import (
	"net/http"
	"stock/internal/model"
	"stock/internal/repository"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// PermissionMiddleware 权限验证中间件
type PermissionMiddleware struct {
	db       *gorm.DB
	userRepo *repository.UserRepository
	roleRepo *repository.RoleRepository
}

// NewPermissionMiddleware 创建权限中间件
func NewPermissionMiddleware(db *gorm.DB) *PermissionMiddleware {
	return &PermissionMiddleware{
		db:       db,
		userRepo: repository.NewUserRepository(db),
		roleRepo: repository.NewRoleRepository(db),
	}
}

// RequirePermission 权限验证中间件
func (m *PermissionMiddleware) RequirePermission(permissionCode string) gin.HandlerFunc {
	return func(c *gin.Context) {
		// 获取用户信息
		userID, exists := c.Get("userID")
		if !exists {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "未认证用户",
			})
			c.Abort()
			return
		}

		user, err := m.userRepo.GetByID(userID.(uint))
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "用户信息获取失败",
			})
			c.Abort()
			return
		}

		// 检查用户是否拥有权限
		if !m.hasPermission(user, permissionCode) {
			c.JSON(http.StatusForbidden, gin.H{
				"code":    403,
				"message": "权限不足",
			})
			c.Abort()
			return
		}

		c.Next()
	}
}

// RequireAnyPermission 检查用户是否拥有任意一个权限
func (m *PermissionMiddleware) RequireAnyPermission(permissionCodes ...string) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID, exists := c.Get("userID")
		if !exists {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "未认证用户",
			})
			c.Abort()
			return
		}

		user, err := m.userRepo.GetByID(userID.(uint))
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "用户信息获取失败",
			})
			c.Abort()
			return
		}

		// 检查用户是否拥有任意一个权限
		for _, permissionCode := range permissionCodes {
			if m.hasPermission(user, permissionCode) {
				c.Next()
				return
			}
		}

		c.JSON(http.StatusForbidden, gin.H{
			"code":    403,
			"message": "权限不足",
		})
		c.Abort()
	}
}

// hasPermission 检查用户是否拥有指定权限
func (m *PermissionMiddleware) hasPermission(user *model.User, permissionCode string) bool {
	// 检查用户是否有角色
	if user.RoleID == nil || *user.RoleID == 0 {
		return false
	}

	// 获取角色信息
	role, err := m.roleRepo.GetByID(*user.RoleID)
	if err != nil || role.Status != 1 {
		return false
	}

	// 如果是管理员，拥有所有权限
	if role.RoleCode == "admin" || role.RoleCode == "super_admin" {
		return true
	}

	// 获取角色权限
	permissions, err := m.roleRepo.GetRolePermissions(*user.RoleID)
	if err != nil {
		return false
	}

	// 检查用户角色权限
	for _, permission := range permissions {
		if permission.PermissionCode == permissionCode && permission.Status == 1 {
			return true
		}
	}

	return false
}

// GetUserPermissions 获取用户权限列表
func (m *PermissionMiddleware) GetUserPermissions(userID uint) ([]string, error) {
	user, err := m.userRepo.GetByID(userID)
	if err != nil {
		return nil, err
	}

	var permissions []string
	if user.RoleID != nil {
		// 通过角色仓储获取角色权限
		rolePermissions, err := m.roleRepo.GetRolePermissions(*user.RoleID)
		if err != nil {
			return nil, err
		}

		for _, permission := range rolePermissions {
			if permission.Status == model.PermissionStatusActive {
				permissions = append(permissions, permission.PermissionCode)
			}
		}
	}

	return permissions, nil
}

// GetUserMenuPermissions 获取用户菜单权限
func (m *PermissionMiddleware) GetUserMenuPermissions(userID uint) ([]model.Permission, error) {
	user, err := m.userRepo.GetByID(userID)
	if err != nil {
		return nil, err
	}

	var menuPermissions []model.Permission
	if user.RoleID != nil && *user.RoleID > 0 {
		// 获取角色权限
		permissions, err := m.roleRepo.GetRolePermissions(*user.RoleID)
		if err != nil {
			return nil, err
		}

		for _, permission := range permissions {
			// 只返回菜单类型的权限
			if permission.Status == model.PermissionStatusActive &&
				permission.ResourceType != nil &&
				(*permission.ResourceType == "menu" || *permission.ResourceType == "module") {
				menuPermissions = append(menuPermissions, permission)
			}
		}
	}

	return menuPermissions, nil
}

// RoutePermissionMap 路由权限映射
var RoutePermissionMap = map[string]string{
	// 股票数据
	"/stocks":           "stocks:list",
	"/stocks/watchlist": "stocks:watchlist",
	"/stocks/realtime":  "stocks:realtime",

	// 策略管理
	"/strategies":          "strategies:list",
	"/strategies/create":   "strategies:create",
	"/strategies/backtest": "strategies:backtest",

	// 选股结果
	"/selections":         "selections:list",
	"/selections/today":   "selections:today",
	"/selections/history": "selections:history",

	// 投资组合
	"/portfolios":             "portfolios:manage",
	"/portfolios/performance": "portfolios:performance",

	// 数据采集
	"/collectors":       "collectors:manage",
	"/collectors/tasks": "collectors:tasks",
	"/collectors/sync":  "collectors:sync",

	// 通知管理
	"/notifications/robots":    "notifications:robots",
	"/notifications/templates": "notifications:templates",
	"/notifications/logs":      "notifications:logs",

	// 报表分析
	"/reports/strategy": "reports:strategy",
	"/reports/market":   "reports:market",
	"/reports/risk":     "reports:risk",

	// 系统管理
	"/system/users":      "system:users",
	"/system/roles":      "system:roles",
	"/system/monitoring": "system:monitoring",
	"/system/settings":   "system:settings",
}

// RequireRoutePermission 根据路由自动检查权限
func (m *PermissionMiddleware) RequireRoutePermission() gin.HandlerFunc {
	return func(c *gin.Context) {
		path := c.Request.URL.Path

		// 查找对应的权限编码
		permissionCode, exists := RoutePermissionMap[path]
		if !exists {
			// 如果路由没有配置权限要求，直接通过
			c.Next()
			return
		}

		// 获取用户信息
		userID, exists := c.Get("userID")
		if !exists {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "未认证用户",
			})
			c.Abort()
			return
		}

		user, err := m.userRepo.GetByID(userID.(uint))
		if err != nil {
			c.JSON(http.StatusUnauthorized, gin.H{
				"code":    401,
				"message": "用户信息获取失败",
			})
			c.Abort()
			return
		}

		// 检查权限
		if !m.hasPermission(user, permissionCode) {
			c.JSON(http.StatusForbidden, gin.H{
				"code":    403,
				"message": "权限不足",
			})
			c.Abort()
			return
		}

		c.Next()
	}
}
